components/uniset/src/uniset.rs

```diff
@@ -183,6 +188,93 @@ impl UnicodeSet {
             None => false,
         }
     }
+
+    /// Check if the calling Unicodeset contains all the characters of the given UnicodeSet
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSet;
+    /// let example_list = vec![65, 70, 85, 91]; // A - E, U - Z
+    /// let example = UnicodeSet::from_inversion_list(example_list).unwrap();
+    /// let a_to_d = UnicodeSet::from_inversion_list(vec![65, 69]).unwrap();
+    /// let f_to_t = UnicodeSet::from_inversion_list(vec![70, 85]).unwrap();
+    /// let r_to_x = UnicodeSet::from_inversion_list(vec![82, 88]).unwrap();
+    /// assert!(example.contains_set(&a_to_d)); // contains all
+    /// assert!(!example.contains_set(&f_to_t)); // contains none
+    /// assert!(!example.contains_set(&r_to_x)); // contains some
+    /// ```
+    pub fn contains_set(&self, set: &UnicodeSet) -> bool {
+        if set.as_inversion_list().len() > self.inv_list.len() {
+            return false;
+        }
+        let s = set.as_inversion_list().len() as f32;
+        let u = self.inv_list.len() as f32;
+        if s * u.log2() < u {
+            set.inv_list.chunks(2).all(|range| {
+                self.contains_range(
+                    &(char::from_u32(range[0]).unwrap()..char::from_u32(range[1]).unwrap()),
+                )
+            })
+        } else {
+            let mut set_ranges: Chunks<u32> = set.as_inversion_list().chunks(2);
+            let mut check = set_ranges.next();
+            for range in self.inv_list.chunks(2) {
+                match check {
+                    Some(r) => {
+                        if r[0] >= range[0] && r[1] <= range[1] {
+                            check = set_ranges.next();
+                        }
+                    }
+                    _ => break,
+                }
+            }
+            check.is_none()
+        }
+    }
+
+    /// Returns the end of the initial substring where the characters are either contained/not contained
+    /// in the set.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use icu_unicodeset::UnicodeSet;
+    /// let example_list = vec![65, 68]; // {A, B, C}
+    /// let example = UnicodeSet::from_inversion_list(example_list).unwrap();
+    /// assert_eq!(example.span("CABXYZ", true), 3);
+    /// assert_eq!(example.span("XYZC", false), 3);
+    /// assert_eq!(example.span("XYZ", true), 0);
+    /// assert_eq!(example.span("ABC", false), 0);
+    /// ```
+    pub fn span(&self, span_str: &str, contained: bool) -> usize {
+        span_str
+            .chars()
+            .take_while(|&x| self.contains(x) == contained)
+            .count()
```

Comment:
@markusicu Can you take a look at this algorithm for `span` (and `span_back` below)?

Reply:
I abstracted the `span` and `span_back` algorithms from ICU4C (but please provide feedback if something seems off):

**ICU4C span** (span_back is just the condition flipped to go backwards)
```c
UChar32 c;
int32_t start = 0, prev = 0;
do {
    U16_NEXT(str, start, length, c);
    if (spanCondition  != contains(c)) {
        break;
    }
} while((prev = start) < length);
return prev
```

**ICU4X span**
```rust
span_str.chars().take_while(|&x| self.contains(x) == contained).count()
```